pkg/ddevapp/commands.go

```diff
@@ -1,47 +1,48 @@
 package ddevapp
 
 import (
-	"github.com/ddev/ddev/pkg/fileutil"
-	"github.com/ddev/ddev/pkg/globalconfig"
-	"github.com/ddev/ddev/pkg/util"
-	copy2 "github.com/otiai10/copy"
+	"fmt"
 	"os"
+	"path"
 	"path/filepath"
-)
 
-// PopulateCustomCommandFiles sets up the custom command files in the project
-// directories where they need to go.
-func PopulateCustomCommandFiles(app *DdevApp) error {
+	dockerImages "github.com/ddev/ddev/pkg/docker"
+	"github.com/ddev/ddev/pkg/dockerutil"
+	"github.com/ddev/ddev/pkg/globalconfig"
+	"github.com/ddev/ddev/pkg/nodeps"
+	"github.com/ddev/ddev/pkg/util"
+)
 
+// PopulateGlobalCustomCommandFiles sets up the global command files
+// ddev-global-cache where they belong
+func PopulateGlobalCustomCommandFiles() error {
```

Comment:
In this function we do **three** different container runs, so the `ddev start` performance impact is nontrivial. 

* RunSimpleContainer()
* CopyIntoVolume()
* RunSimpleContainer() again

Some thoughts about mitigation would be to 
1. Do this work after the webserver is started, where we have ddev-global-cache already mounted? Then it's just one app.Exec() maybe?
2. Consolidate it into a script so we only have one container run?

I think the idea is all fine, I just don't want to add more container runs to get us through the `ddev start`

Reply:
> Do this work after the webserver is started, where we have ddev-global-cache already mounted? Then it's just one app.Exec() maybe?

Can you please explain how that would cut down the number of executions? We still need to do 3 things;
1. Remove the old contents. This is necessary to avoid keeping old commands that have since been removed from `~/.ddev/commands`
1. Copy commands into the volume.
1. `chmod` the files so commands and completion scripts can be executed.

I don't know how to copy the commands into the volume without using `dockerutil.CopyIntoVolume()`. If you can show me a way to do that which can be combined with the first and third steps above, I'd be happy to try that.

Reply:
I guess an alternative approach would be to bind `~/.ddev/commands` directly as a volume for all containers - then we only need the `chmod` and we can do that on the host.
I don't know what flow-on effects that might have though (I don't know much about docker volumes), and it's definitely a larger departure from how these are currently handled.

Reply:
Ahh I've been mulling this over - is the problem more about the fact that anonymous containers are being started for 1 and 3 above, which is slower than just a `docker exec` in an existing container?

If that's the problem, I can swap those out for running those commands in the web container if it's available (I say "if it's available" because remember this is for `ddev debug fix-commands` as well, which might run when the project is stopped)

Reply:
> is the problem more about the fact that anonymous containers are being started for 1 and 3 above

Yes, exactly. The Exec() doesn't take much time, but running a new container does. You're on the right track. It's just that we already have too many of those CopyIntoVolume() calls in `ddev start`. 

Since `ddev debug fix-commands` is a one-off it can be done any way we want. It would be sad for it to have different code, perhaps an if-started/else would do the job?

Reply:
> It's just that we already have too many of those CopyIntoVolume() calls in ddev start.

As I mentioned earlier I'm not sure I can do anything about the `CopyIntoVolume()` call, but I can add conditionals to use the web container for the other two if it's available, so I'll do that.

If you can think of a way to use the web container to avoid calling `CopyIntoVolume()`, please let me know.

Reply:
There we go - if there's an active running app we use `app.Exec()` but otherwise (e.g. if running from outside a project dir) it uses the anonymous container.

Reply:
Looks like it did actually run after starting the containers during `ddev start`.
I've changed the order now so that the containers are set up and ready first. See https://github.com/ddev/ddev/compare/6ac57710c75a3840425def8178aa333ab25941b8..fec7914a43a1b7c758d721efbf28a543357ec9ae

The post-start hook runs *after* this, so I think there is no effective change for users.